File: LC/[0015] 3Sum.cpp
/*
* 20240526
*/

class Solution {

    vector<vector<int>> two_sum(int st, int target, vector<int> & nums){
        int ll=st, rr=nums.size()-1;
        vector<vector<int>> res;
        while(ll<rr){
            int a=nums[ll], b=nums[rr];
            if(a+b==target){       
                res.push_back({a, b});
                ++ll;
                --rr;
                while(ll>st && ll < rr && nums[ll]==nums[ll-1])++ll;
            }else if(a+b > target){
                --rr;
            }else{
                ++ll;
            }
        }
        return res;
    }

public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        int N=nums.size();
        
        vector<vector<int>> res;
        for(int ii=0; ii<N-2; ++ii){
            if(ii>0 && nums[ii]==nums[ii-1]) continue;
            vector<vector<int>> twos = two_sum(ii+1, 0-nums[ii], nums);

            for(int jj=0; jj<twos.size(); ++jj){
                res.push_back({nums[ii], twos[jj][0], twos[jj][1]});
            }
        }

        return res;        
    }
};

/* --------------------------------------------------- */
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        sort(nums.begin(), nums.end());
        int N=nums.size();
        vector<vector<int>> res;
        for(int ii=0; ii<N; ++ii){
            if(nums[ii] > 0) break;
            if(ii==0 || nums[ii] != nums[ii-1])//O(N) time
                two_sum_sorted(nums, ii, res);
        }
        
        return res;
    }

    void two_sum_sorted(vector<int> & nums, int idx, vector<vector<int>> & res){
        int ll=idx+1, rr=nums.size()-1;
        while(ll<rr){
            int sum=nums[idx]+nums[ll]+nums[rr];

            if(sum==0){
                res.push_back({nums[idx], nums[ll], nums[rr]});
                ++ll;
                while(ll<rr and nums[ll]==nums[ll-1])++ll;
                // rr-=1
            }else if(sum > 0){
                --rr;
            }else{
                ++ll;
            }
        }
    }
};

File: LC/[0015] 3Sum.java
//20240827
class Solution {
    private List<Pair<Integer, Integer>> twoSum(int[] nums, int st, int target){
        int ll=st, rr=nums.length-1;
        List<Pair<Integer, Integer>> ans = new ArrayList<>();
        while(ll<rr){
            while(ll>st && ll< rr && nums[ll]==nums[ll-1]){//avoid duplicates
                ++ll;
                continue;
            }
            if(ll < rr && nums[ll]+nums[rr]==target){
                ans.add(new Pair(nums[ll], nums[rr]));
                ++ll;
                --rr;
            }else if(nums[ll]+nums[rr]<=target){
                ++ll;
            }else{
                --rr;
            }
        }

        return ans;
    }
    public List<List<Integer>> threeSum(int[] nums) {
        int N = nums.length;
        Arrays.sort(nums);
        List<List<Integer>> ans = new ArrayList<>();
        
        for(int ii=0; ii<N-2; ++ii){//[0, N-3], T=N, need three
            if(ii>0 && nums[ii]==nums[ii-1]) continue;
            List<Pair<Integer, Integer>> twoSumAns = twoSum(nums, ii+1, -nums[ii]);
            for(var pair: twoSumAns){
                ans.add(List.of(nums[ii], pair.getKey(), pair.getValue()));
            }
        }      
        return ans;
    }
}
//----------
class Solution {
    public List<List<Integer>> two_sum(int st, int target, int[] nums){
        List<List<Integer>> res = new ArrayList<>();//how to declare ArrayList

        int ll=st, rr=nums.length-1; //array length is an attribute
        while(ll<rr){
            int a=nums[ll], b=nums[rr];

            if(a+b==target){
                res.add(Arrays.asList(a, b));//how to construct array from int
                ++ll;
                --rr;
                while(ll<rr && nums[ll]==nums[ll-1])++ll;
            }else if(a+b > target){
                --rr;
            }else{
                ++ll;
            }

        }

        return res;
    }

    public List<List<Integer>> threeSum(int[] nums) {
        Arrays.sort(nums); //how to sort Array
        List<List<Integer>> res = new ArrayList<>(); //Declare ArrayList
        int N = nums.length;//length of Array is an attribute

        for(int ii=0; ii<N-2; ++ii){
            List<List<Integer>> twos = two_sum(ii+1, 0-nums[ii], nums);
            if(ii>0 && nums[ii]==nums[ii-1]) continue;

            for(int jj=0; jj<twos.size(); ++jj){//length of ArrayList is Size
                //how to append number to ArrayList
                res.add(Arrays.asList(nums[ii], twos.get(jj).get(0), twos.get(jj).get(1))); //random access ArrayList          
            }
        }

        return res;

    }
}